### PR TITLE
Balay/petsc remove boost dependency

### PR DESCRIPTION
**Pull request by balay**
ref: #8296

### CODE REVIEW

**Comment by balay** [comment]
```
==================================== ERRORS ====================================
________ ERROR at teardown of test_build_systems[url_and_build_system5] ________

request = <SubRequest 'check_for_leftover_stage_files' for <Function 'test_build_systems[url_and_build_system5]'>>
mock_stage = '/tmp/pytest-of-travis/pytest-0/mock_stage0'
_ignore_stage_files = set(['.lock', 'spack-stage-m58e8djl'])
```
I do not understand this error. @davydden, could you take a look at it?

**Comment by balay** [comment]
@davydden I pushed one more change. Could you review that aswell? Thanks!

**Comment by balay** [comment]
Is this ready for a merge?

cc: @adamjstewart  @tgamblin

File: var/spack/repos/builtin/packages/petsc/package.py [hunk 1]
```diff
@@ -121,14 +120,15 @@ class Petsc(Package):
...
-    depends_on('boost', when='@:3.5+boost')
     depends_on('metis@5:~int64+real64', when='@:3.7.99+metis~int64+double')
     depends_on('metis@5:~int64', when='@:3.7.99+metis~int64~double')
     depends_on('metis@5:+int64+real64', when='@:3.7.99+metis+int64+double')
     depends_on('metis@5:+int64', when='@:3.7.99+metis+int64~double')
     # petsc-3.8+ uses default (float) metis with any (petsc) precision
     depends_on('metis@5:~int64', when='@3.8:+metis~int64')
     depends_on('metis@5:+int64', when='@3.8:+metis+int64')
+    # trilinos might require boost
+    depends_on('boost', when='@3.7.0:+trilinos+boost')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/petsc/package.py]: If that's the case, shouldn't trilinos depend on boost, not petsc?

Comment (balay) [reply var/spack/repos/builtin/packages/petsc/package.py]: Hm - ok - many these lines are not needed.

The important thing here is the next set of changes

```
         if 'trilinos' in spec:
             options.append('--with-cxx-dialect=C++11')
-            options.append('--with-boost=1')
+            if spec.satisfies('^trilinos+boost'):
+                options.append('--with-boost=1')
```

Will update PR

**Comment by balay** [comment]
@adamjstewart  is this PR also ready for merge?

**Pull request merged**